test/shared/src/main/scala/zio/test/mock/MockRandom.scala

```diff
@@ -26,146 +30,141 @@ trait MockRandom extends Random {
-
-    val clearChars: UIO[Unit] =
-      randomState.update(data => data.copy(chars = List.empty)).unit
-
-    val clearStrings: UIO[Unit] =
-      randomState.update(data => data.copy(strings = List.empty)).unit
-
-    val clearBytes: UIO[Unit] =
-      randomState.update(data => data.copy(bytes = List.empty)).unit
-
-    private def nextRandom[T](shift: Data => (T, Data)) =
+    val nextDouble: UIO[Double] =
       for {
-        data            <- randomState.get
-        (next, shifted) = shift(data)
-        _               <- randomState.update(_ => shifted)
-      } yield next
-
-    private def shiftBooleans(data: Data) =
-      (
-        data.booleans.headOption.fold(MockRandom.defaultBoolean)(identity),
-        data.copy(booleans = shiftLeft(data.booleans))
-      )
-
-    private def shiftIntegers(data: Data) =
-      (
-        data.integers.headOption.fold(MockRandom.defaultInteger)(identity),
-        data.copy(integers = shiftLeft(data.integers))
-      )
-
-    private def shiftIntWithLimit(limit: Int)(data: Data) = {
-      val next = data.integers.headOption.fold(MockRandom.defaultInteger)(identity)
-      (Math.min(limit, next), data.copy(integers = shiftLeft(data.integers)))
+        i1 <- next(26)
+        i2 <- next(27)
+      } yield ((i1.toLong << 27) + i2) / (1L << 53).toDouble
+
+    def nextBytes(length: Int): UIO[Chunk[Byte]] = {
+      //  Our RNG generates 32 bit integers so to maximize efficieny we want to
+      //  pull 8 bit bytes from the current integer until it is exhausted
+      //  before generating another random integer
+      def loop(i: Int, rnd: UIO[Int], n: Int, acc: UIO[List[Byte]]): UIO[List[Byte]] =
+        if (i == length)
+          acc.map(_.reverse)
+        else if (n > 0)
+          rnd.flatMap(rnd => loop(i + 1, UIO.succeed(rnd >> 8), n - 1, acc.map(rnd.toByte :: _)))
+        else
+          loop(i, nextInt, (length - i) min 4, acc)
+
+      loop(0, nextInt, length min 4, UIO.succeed(List.empty[Byte])).map(Chunk(_: _*))
     }
 
-    private def shiftDoubles(data: Data) =
-      (data.doubles.headOption.fold(MockRandom.defaultDouble)(identity), data.copy(doubles = shiftLeft(data.doubles)))
+    val nextFloat: UIO[Float] =
+      next(24).map(_ / (1 << 24).toFloat)
+
+    val nextGaussian: UIO[Double] =
+      //  The Box-Muller transform generates two normally distributed random
+      //  doubles, so we store the second double in a queue and check the
+      //  queue before computing a new pair of values to avoid wasted work.
+      randomState.modify {
+        case Data(seed, queue) =>
+          queue.dequeueOption.fold { (Option.empty[Double], Data(seed, queue)) } {
+            case (d, queue) => (Some(d), Data(seed, queue))
+          }
+      }.flatMap {
+        case Some(nextNextGaussian) => UIO.succeed(nextNextGaussian)
+        case None =>
+          def loop: UIO[(Double, Double, Double)] =
+            nextDouble.flatMap { d1 =>
+              nextDouble.flatMap { d2 =>
```

Comment:
`(nextDouble zip nextDouble)` can be handy for this pattern.

Reply:
Yes!